File: aspnet-in-practice/06/CS/CustomControls.Databinding/SingleView.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.ComponentModel;
using System.Web.UI.WebControls;
using System.Collections;
using System.Web.UI;

namespace CustomControls.Databinding
{
	[ParseChildren(true)]
	public class SingleView : CompositeDataBoundControl, INamingContainer
	{
		[Bindable(true)]
		public override object DataSource { get; set; }

		[Browsable(false), PersistenceMode(PersistenceMode.InnerProperty), DefaultValue(typeof(ITemplate), ""), TemplateContainer(typeof(RepeaterItem), BindingDirection.TwoWay)]
		public ITemplate ItemTemplate { get; set; }

		[Browsable(false), PersistenceMode(PersistenceMode.InnerProperty), TemplateContainer(typeof(RepeaterItem)), DefaultValue((string)null)]
		public ITemplate EmptyTemplate { get; set; }

		// this methods will prevent any container tag
		public override void RenderBeginTag(HtmlTextWriter writer)
		{
		}

		public override void RenderEndTag(HtmlTextWriter writer)
		{
		}
		// ***********

		protected override int CreateChildControls(IEnumerable dataSource, bool dataBinding)
		{
			if (ItemTemplate == null)
				throw new ArgumentNullException("ItemTemplate");

			// we need a control container
			RepeaterItem container = new RepeaterItem(0, ListItemType.Item);
			this.Controls.Add(container);
			ItemTemplate.InstantiateIn(container);

			// if databinding is requested, perform it
			if (dataBinding)
			{
				// check for DataSource and EmptyTemplate
				if (DataSource == null)
				{
					if (EmptyTemplate != null)
					{
						this.Controls.Clear();
						EmptyTemplate.InstantiateIn(this);
					}
				}
				else
				{
					container.DataItem = DataSource;
					if (!Page.IsPostBack)
						container.DataBind();
					container.DataItem = null;
				}
			}

			return 1;
		}

		
	}

}

File: aspnet-in-practice/06/VB/CustomControls.Web/CustomControls.Databinding/SingleView.vb
Imports System.Collections.Generic
Imports System.Linq
Imports System.Text
Imports System.ComponentModel
Imports System.Web.UI.WebControls
Imports System.Collections
Imports System.Web.UI

<ParseChildren(True)>
Public Class SingleView
	Inherits CompositeDataBoundControl
	Implements INamingContainer
	<Bindable(True)>
	Public Overrides Property DataSource() As Object

	<Browsable(False), PersistenceMode(PersistenceMode.InnerProperty), DefaultValue(GetType(ITemplate), ""), TemplateContainer(GetType(RepeaterItem), BindingDirection.TwoWay)>
	Public Property ItemTemplate() As ITemplate

	<Browsable(False), PersistenceMode(PersistenceMode.InnerProperty), TemplateContainer(GetType(RepeaterItem)), DefaultValue(DirectCast(Nothing, String))>
	Public Property EmptyTemplate() As ITemplate

	' this methods will prevent any container tag
	Public Overrides Sub RenderBeginTag(ByVal writer As HtmlTextWriter)
	End Sub

	Public Overrides Sub RenderEndTag(ByVal writer As HtmlTextWriter)
	End Sub
	' ***********

	Protected Overrides Function CreateChildControls(ByVal dataSource__1 As IEnumerable, ByVal dataBinding As Boolean) As Integer
		If ItemTemplate Is Nothing Then
			Throw New ArgumentNullException("ItemTemplate")
		End If

		' we need a control container
		Dim container As New RepeaterItem(0, ListItemType.Item)
		Me.Controls.Add(container)
		ItemTemplate.InstantiateIn(container)

		' if databinding is requested, perform it
		If dataBinding Then
			' check for DataSource and EmptyTemplate
			If DataSource Is Nothing Then
				If EmptyTemplate IsNot Nothing Then
					Me.Controls.Clear()
					EmptyTemplate.InstantiateIn(Me)
				End If
			Else
				container.DataItem = DataSource
				If Not Page.IsPostBack Then
					container.DataBind()
				End If
				container.DataItem = Nothing
			End If
		End If

		Return 1
	End Function


End Class